File: csharp/TetrisGameTests/PeriodicTaskTest.cs
using hu.klenium.tetris.Util;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using TetrisGameTests.Helpers;

namespace TetrisGameTests
{
    [TestClass]
    public class PeriodicTaskTest
    {
        private static readonly int PERIOD = 500;
        private static readonly int OFFSET = PERIOD / 2;
        private PeriodicTask clock = null;
        private int foo = 1;
        [TestInitialize] public void SetUp()
        {
            clock = new PeriodicTask(() => ++foo, PERIOD);
        }
        [TestMethod] public void NonStarted()
        {
            TestUtil.RunLater(OFFSET + PERIOD, () => Assert.AreEqual(foo, 1));
        }
        [TestMethod] public void MultiplePeriods()
        {
            clock.Start();
            TestUtil.RunLater(OFFSET, () => Assert.AreEqual(foo, 1));
            TestUtil.RunLater(PERIOD, () => Assert.AreEqual(foo, 2));
            TestUtil.RunLater(PERIOD, () => Assert.AreEqual(foo, 3));
        }
        [TestMethod] public void Stoped()
        {
            clock.Start();
            TestUtil.RunLater(OFFSET, () => {
                Assert.AreEqual(foo, 1);
                clock.Stop();
            });
            TestUtil.RunLater(PERIOD, () => Assert.AreEqual(foo, 1));
            TestUtil.RunLater(PERIOD, () => Assert.AreEqual(foo, 1));
        }
        [TestMethod] public void ResetPeriodTime()
        {
            clock.Start();
            TestUtil.RunLater(OFFSET, () => {
                Assert.AreEqual(foo, 1);
                clock.Reset();
            });
            TestUtil.RunLater(OFFSET, () => Assert.AreEqual(foo, 1));
            TestUtil.RunLater(PERIOD, () => Assert.AreEqual(foo, 2));
        }
    }
}

File: python/tests/PeriodicTask_test.py
import unittest
from tests.helpers import util
from klenium.tetris.util.PeriodicTask import PeriodicTask
from tests.helpers.time_constant import PERIOD as PERIOD
from tests.helpers.time_constant import OFFSET as OFFSET


class PeriodicTaskSlowTest(unittest.TestCase):
    def setUp(self):
        self.foo = 1
        self.clock = PeriodicTask(lambda: ++self.foo, PERIOD)

    def test_non_started(self):
        util.run_later(OFFSET + PERIOD, lambda: self.assertEqual(self.foo, 1))

    def test_multiple_periods(self):
        self.clock.start()
        util.run_later(OFFSET, lambda: self.assertEqual(self.foo, 1))
        util.run_later(PERIOD, lambda: self.assertEqual(self.foo, 2))
        util.run_later(PERIOD, lambda: self.assertEqual(self.foo, 3))

    def test_stoped(self):
        self.clock.start()
        util.run_later(OFFSET, lambda: (
            self.assertEqual(self.foo, 1),
            self.clock.stop()
        ))
        util.run_later(PERIOD, lambda: self.assertEqual(self.foo, 1))
        util.run_later(PERIOD, lambda: self.assertEqual(self.foo, 1))

    def test_reset_period_time(self):
        self.clock.start()
        util.run_later(OFFSET, lambda: (
            self.assertEqual(self.foo, 1),
            self.clock.reset()
        ))
        util.run_later(OFFSET, lambda: self.assertEqual(self.foo, 1))
        util.run_later(PERIOD, lambda: self.assertEqual(self.foo, 2))
